### PR TITLE
Configure pre-commit hooks for the whole project

### PR DESCRIPTION
**Pull request by Benjaminimal**
The main purpose of this is that the whole project has a coherent style which makes reading and especially reviewing easier.
I used [pre-commit](https://pre-commit.com/) to ensure that:
- Files are formatted the same way.
- Frontend code is linted.

Since the two tools didn't work well with each other I removed husky and replaced what it was doing with pre-commit.
While doing that I removed the build frontend step from the pre-commit hooks as this just takes too long and the CI should take care of that.

## Basics

- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [x] I fixed all affected decisions
- [x] I added code comments, logging, and assertions as appropriate
- [x] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      - id: end-of-file-fixer
+      - id: mixed-line-ending
+      - id: trailing-whitespace
+        args: ["--markdown-linebreak-ext=md"]
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: v2.7.1
+    hooks:
+      - id: prettier
+        name: doc format
+        files: ^doc/
```
Comment (markus2330) [inline .pre-commit-config.yaml:17]: This would exclude e.g. backend/README.md, won't it?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          - eslint-plugin-react-hooks@4.6.0
+  # Backend hooks
+  - repo: https://github.com/doublify/pre-commit-rust
+    rev: v1.0
+    hooks:
+      - id: fmt
+        name: backend format
+        args: [--manifest-path=backend/Cargo.toml]
+        files: ^backend/
+        pass_filenames: false
```
Comment (markus2330) [inline .pre-commit-config.yaml:64]: Please give a bit of information about which config you changed (and why?).

File: doc/contrib/README.md [hunk 3]
```diff
@@ -12,6 +12,18 @@ If you find a bug in the code or a mistake in the documentation, you can help us
 
 ## Development
 
+### Hooks
```
Comment (markus2330) [inline doc/contrib/README.md:15]: We should either mention hooks in .github/CONTRIBUTING.md or even better: link [doc/contrib/README.md](https://github.com/ElektraInitiative/PermaplanT/pull/160/files#diff-d5f5656eaeb7cfba734b6f3046c390588e849c59f79f74d1de13067af6a92f58) from there.

File: frontend/package-lock.json
Comment (markus2330) [inline frontend/package-lock.json:1]: This seems to be an unrelated change to the rest of the PR (better to avoid something like this in general unless it is small documentation fixes).  What are the pros/cons of having this included or was it only a mistake? For the backend, we definitely should have the lock file within the repository.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Great job, I like pre-commit, it works really well. There are some open points in this PR, though. Anyway, we shouldn't merge it unless we didn't reformat everything.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Benjaminimal) [reply .pre-commit-config.yaml:17]: True.
It would omit the root README.md as well.
I will adjust it.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (Benjaminimal) [reply .pre-commit-config.yaml:64]: I'm not sure what you mean.
I didn't change any config for the backend formatting.
All I did was tell pre-commit how to call `cargo fmt` on the backend subfolder only.

File: frontend/package-lock.json
Comment (Benjaminimal) [reply frontend/package-lock.json:1]: It's not unrelated.
I uninstalled husky and lint-staged which made changes to the lock file.

Comment (Benjaminimal) [reply frontend/package-lock.json:1]: According to npm this file should be checked into vcs.

> This file is intended to be committed into source repositories

https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json

File: .pre-commit-config.yaml (hunk 2 above)
Comment (markus2330) [reply .pre-commit-config.yaml:64]: I mean that you explain e.g. what `pass_filenames` is and does (or why it is false here) as inline-comment.

File: frontend/package-lock.json
Comment (markus2330) [reply frontend/package-lock.json:1]: Ok, thx! Then it is okay to have it here.

File: doc/contrib/README.md (hunk 3 above)
Comment (Benjaminimal) [reply doc/contrib/README.md:15]: see
https://github.com/ElektraInitiative/PermaplanT/pull/160/commits/38e26b095b5ff3f665af19ec9d4a5a2078e254bc
https://github.com/ElektraInitiative/PermaplanT/pull/160/commits/6c983bb656c708b0f1f51a5ab049c6819d55b9a9

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Benjaminimal) [reply .pre-commit-config.yaml]: see
https://github.com/ElektraInitiative/PermaplanT/pull/160/commits/96b260a9ffc2290f8fda6009241f118b3a5f19d2

File: .pre-commit-config.yaml (hunk 2 above)
Comment (Benjaminimal) [reply .pre-commit-config.yaml:64]: Sure.
1. I need to pass `args: [--manifest-path=backend/Cargo.toml]` because cargo is called from the root project dir which doesn't include `Cargo.toml` for the backend.
2. The `files: ^backend/` option tells pre-commit to skip this hook if no files in the backend dir is staged.
3. `pass_filenames: falsepass_filenames: false` tells pre-commit not to pass the staged & matched file names to the executed command, `cargo fmt`. I did this because the command doesn't take any file name arguments. It would complain and not run properly without it.

**Comment by Benjaminimal** [comment]
> Anyway, we shouldn't merge it unless we didn't reformat everything.

Sure I can rebase on master and run hooks on the whole project.

Just tell me when so that it doesn't stay open for too long and gets stale again due to new things getting merged.

**Review by markus2330 — APPROVED** [review]
I think we can merge this now, can you rebase and apply the reformatting?

**Comment by Benjaminimal** [comment]
@markus2330 
I rebased and applied all the hooks.
Sadly the frontend linting errors have increased.
```
frontend/src/features/seeds/routes/SeedDetails.tsx
  36:6  warning  React Hook useEffect has a missing dependency: 'id'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

✖ 1 problem (0 errors, 1 warning)

ESLint found too many warnings (maximum: 0).

frontend/src/components/Form/CreatableSelectMenu.tsx
  15:34  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  15:62  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

frontend/src/features/seeds/routes/CreateSeed.tsx
  51:6  warning  React Hook useEffect has a missing dependency: 'searchPlants'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

✖ 3 problems (0 errors, 3 warnings)

ESLint found too many warnings (maximum: 0).

frontend/src/config/index.ts
  15:22  error  Do not access Object.prototype method 'hasOwnProperty' from target object  no-prototype-builtins

✖ 1 problem (1 error, 0 warnings)


frontend/src/tailwind.config.js
    1:7   warning  'colors' is assigned a value but never used     @typescript-eslint/no-unused-vars
    1:16  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
    1:16  error    'require' is not defined                        no-undef
   55:7   warning  'brown' is assigned a value but never used      @typescript-eslint/no-unused-vars
  103:7   warning  'blue' is assigned a value but never used       @typescript-eslint/no-unused-vars
  140:1   error    'module' is not defined                         no-undef

frontend/tailwind.config.js
  2:24  error  Require statement not part of import statement  @typescript-eslint/no-var-requires

✖ 7 problems (4 errors, 3 warnings)


frontend/src/features/seeds/components/CreateSeedForm.tsx
  2:8  warning  'CreatableSelectMenu' is defined but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (0 errors, 1 warning)
frontend/src/features/seeds/routes/SeedDetails.tsx
  36:6  warning  React Hook useEffect has a missing dependency: 'id'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

✖ 1 problem (0 errors, 1 warning)

ESLint found too many warnings (maximum: 0).

frontend/src/components/Form/CreatableSelectMenu.tsx
  15:34  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  15:62  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

frontend/src/features/seeds/routes/CreateSeed.tsx
  51:6  warning  React Hook useEffect has a missing dependency: 'searchPlants'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

✖ 3 problems (0 errors, 3 warnings)

ESLint found too many warnings (maximum: 0).

frontend/src/config/index.ts
  15:22  error  Do not access Object.prototype method 'hasOwnProperty' from target object  no-prototype-builtins

✖ 1 problem (1 error, 0 warnings)


frontend/src/tailwind.config.js
    1:7   warning  'colors' is assigned a value but never used     @typescript-eslint/no-unused-vars
    1:16  error    Require statement not part of import statement  @typescript-eslint/no-var-requires
    1:16  error    'require' is not defined                        no-undef
   55:7   warning  'brown' is assigned a value but never used      @typescript-eslint/no-unused-vars
  103:7   warning  'blue' is assigned a value but never used       @typescript-eslint/no-unused-vars
  140:1   error    'module' is not defined                         no-undef

frontend/tailwind.config.js
  2:24  error  Require statement not part of import statement  @typescript-eslint/no-var-requires

✖ 7 problems (4 errors, 3 warnings)


frontend/src/features/seeds/components/CreateSeedForm.tsx
  2:8  warning  'CreatableSelectMenu' is defined but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (0 errors, 1 warning)
```

**Comment by Benjaminimal** [comment]
We can merge this and even leave the linter in place.
Only staged files when committing will be checked by the linter.
So the first person to touch one of those will have to fix it.

**Comment by Benjaminimal** [comment]
Seems like something was merged to master.
Rebased and reformatted again.

**Comment by markus2330** [comment]
Great job!
